### PR TITLE
Impemented an LRU Cache

### PR DESCRIPTION
**Pull request by rodsan0**
A Least-Recently Used Cache is a queue that sorts elements by access time. When a new element enters the cache, the one that was accessed the longest ago leaves it.

### CODE REVIEW

File: tests/tools/QueueCache.cc [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+    // Make sure there are no elements in cache.
+    REQUIRE(qch.Size() == 0);
+
+    // Make sure capacity is set.
+    REQUIRE(qch.Capacity() == capacity);
+
+    WHEN("elements are stored in it") {
+      // Put some things in the cache.
+      auto it_a = qch.Put('a', 0);
```
Comment (mmore500) [inline tests/tools/QueueCache.cc:24]: Const?

File: source/tools/QueueCache.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+				if (!cache_map.count(key)) {
+					// value is not in cache, so we call function
+					Put(key, fun(key));
+				}
+				// value is in cache, so we find it
+				return UpdateCache(cache_map.find(key)->second);
+			}
+
+			/// Resizes the cache.
+			void SetCapacity(size_t _capacity) {
```
Comment (mmore500) [inline source/tools/QueueCache.h:129]: Const arg

File: source/tools/QueueCache.h [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+			/// Returns true if cache has key.
+			bool Contains(const Key& key) {
+				return cache_map.count(key);
+			}
+
+			/// Stores element in front of cache.
+			typename cache_t::iterator Put(const Key& key, const Value& val) {
+				// try to find element in map
+				auto found = cache_map.find(key);
```
Comment (mmore500) [inline source/tools/QueueCache.h:98]: Const?

File: source/tools/QueueCache.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+			/// Clears the cache.
+			void Clear() {
+				cache.clear();
+				cache_map.clear();
+			}
+
+			/// Deletes element from cache.
+			void Delete(const Key& key) {
+				auto it = cache_map.find(key);
```
Comment (mmore500) [inline source/tools/QueueCache.h:86]: const?

File: source/tools/QueueCache.h [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+					cache_map.erase(cache.back().first);
+					cache.pop_back();
+				}
+			}
+
+			static Value DefaultFun(const Key&) {
+				throw std::invalid_argument("Key not in cache.");
+			}
+
+			void Delete(typename cache_map_t::iterator it) {
```
Comment (mmore500) [inline source/tools/QueueCache.h:60]: Const?

Comment (mmore500) [reply source/tools/QueueCache.h:60]: Argument

File: source/tools/QueueCache.h [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+				cache_map.erase(it);
+				cache.erase(it->second);
+			}
+
+		public:
+			QueueCache(size_t _capacity = DefaultCapacity) : capacity(_capacity) { ; }
```
Comment (mmore500) [inline source/tools/QueueCache.h:66]: Const arg

File: source/tools/QueueCache.h [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+			using cache_t = typename std::list<std::pair<Key, Value>>;
+			using cache_map_t = std::unordered_map<Key, typename cache_t::iterator, Hash, Pred>;
+
+			cache_t cache;
+			cache_map_t cache_map;
+
+
+			size_t capacity;
+
+			Value& UpdateCache(typename cache_t::iterator it) {
```
[the remark this answers is not in the document]
Comment (mmore500) [reply source/tools/QueueCache.h:39]: Const it?

File: source/tools/QueueCache.h [hunk 8]
```diff
@@ -0,0 +1,160 @@
+/**
+ *  @note This file is part of Empirical, https://github.com/devosoft/Empirical
+ *  @copyright Copyright (C) Michigan State University, MIT Software license; see doc/LICENSE.md
+ *  @date 2016-2020
+ *
+ *  @file  QueueCache.h
+ *  @brief
```
Comment (mmore500) [inline source/tools/QueueCache.h:7]: Delete node, add Brief. Set date to 2020

Comment (rodsan0) [reply source/tools/QueueCache.h]: Done in #053f916.

File: source/tools/QueueCache.h (hunk 6 above)
Comment (rodsan0) [reply source/tools/QueueCache.h]: made const in a3fd067

File: source/tools/QueueCache.h (hunk 5 above)
Comment (rodsan0) [reply source/tools/QueueCache.h]: done in a547984

File: source/tools/QueueCache.h (hunk 4 above)
Comment (rodsan0) [reply source/tools/QueueCache.h]: removed variable in c740e0e

File: source/tools/QueueCache.h (hunk 3 above)
Comment (rodsan0) [reply source/tools/QueueCache.h]: done in a3fd067

File: source/tools/QueueCache.h (hunk 2 above)
Comment (rodsan0) [reply source/tools/QueueCache.h]: done in d394a1a

File: source/tools/QueueCache.h [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+				cache_map.clear();
+			}
+
+			/// Deletes element from cache.
+			void Delete(const Key& key) {
+				auto it = cache_map.find(key);
+				Delete(it);
+			}
+
+			/// Returns true if cache has key.
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h]: done in d394a1a

File: source/tools/QueueCache.h [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+			const typename cache_t::const_iterator begin() const { return cbegin(); }
+			const typename cache_t::const_iterator end() const { return cend(); }
+
+			/// Gets an element from cache if found, and creates it otherwise.
+			/// Value type must have a default constructor and an assignment operator.
+			Value& operator[](const Key& index) {
+				// Check whether Value has a default constructor
+				emp_assert(std::is_trivially_constructible<Value>::value);
+
+				// attempt to find key in our iterator map
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h]: changed in eb46264

File: source/tools/QueueCache.h [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+			}
+
+			const typename cache_t::const_iterator begin() const { return cbegin(); }
+			const typename cache_t::const_iterator end() const { return cend(); }
+
+			/// Gets an element from cache if found, and creates it otherwise.
+			/// Value type must have a default constructor and an assignment operator.
+			Value& operator[](const Key& index) {
+				// Check whether Value has a default constructor
+				emp_assert(std::is_trivially_constructible<Value>::value);
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h]: made into a compile-time check in 30f0124

File: source/tools/QueueCache.h [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+		public:
+			QueueCache(size_t _capacity = DefaultCapacity) : capacity(_capacity) { ; }
+			~QueueCache() = default;
+
+			using const_iterator = typename cache_t::const_iterator;
+			using iterator = typename cache_t::const_iterator;
+
+			/// Returns number of elements in cache.
+			size_t Size() { return cache.size(); }
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h]: made const in a547984 and c0c7727

File: source/tools/QueueCache.h [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+			}
+
+			/// Deletes element from cache.
+			void Delete(const Key& key) {
+				auto it = cache_map.find(key);
+				Delete(it);
+			}
+
+			/// Returns true if cache has key.
+			bool Contains(const Key& key) {
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h]: done in c0c7727

File: source/tools/QueueCache.h [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+					Delete(found);
+				}
+
+				// put element into our cache
+				// we use insert because it returns a pointer to our element
+				cache.emplace_front(key, val);
+				// add pointer to this element to our map
+				cache_map.emplace(key, cache.begin());
+
+				Shrink();
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h:126]: explained in 2347aaf

File: source/tools/QueueCache.h [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+			/// Stores element in front of cache.
+			typename cache_t::iterator Put(const Key& key, const Value& val) {
+				// try to find element in map
+				auto found = cache_map.find(key);
+				if (found != cache_map.end()) {
+					Delete(found);
+				}
+
+				// put element into our cache
+				// we use insert because it returns a pointer to our element
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h]: yes, fixed in aa6ab5f

File: source/tools/QueueCache.h [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+			using const_iterator = typename cache_t::const_iterator;
+			using iterator = typename cache_t::const_iterator;
+
+			/// Returns number of elements in cache.
+			size_t Size() { return cache.size(); }
+
+			/// Returns maximum number of elements that will fit in cache.
+			size_t Capacity() { return capacity; }
+
+			/// Clears the cache.
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h]: fixed in d394a1a

File: source/tools/QueueCache.h [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+	>
+	class QueueCache {
+		private:
+			using cache_t = typename std::list<std::pair<Key, Value>>;
+			using cache_map_t = std::unordered_map<Key, typename cache_t::iterator, Hash, Pred>;
+
+			cache_t cache;
+			cache_map_t cache_map;
+
+
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h]: done in ba29095

File: source/tools/QueueCache.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h]: added docstrings in d394a1a
marked const in a547984

File: source/tools/QueueCache.h [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+		size_t DefaultCapacity = std::numeric_limits<size_t>::max(),
+		class Hash = std::hash<Key>,
+		class Pred = std::equal_to<Key>
+	>
+	class QueueCache {
+		private:
+			using cache_t = typename std::list<std::pair<Key, Value>>;
+			using cache_map_t = std::unordered_map<Key, typename cache_t::iterator, Hash, Pred>;
+
+			cache_t cache;
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h]: Closed by ba29095

[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/QueueCache.h]: Solved in a547984 and 7e857e4

File: tests/tools/QueueCache.cc [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+        }
+      );
+
+      THEN("we get the expected result") {
+        REQUIRE(expected == result);
+        CHECK(distance == qch.Size());
+      }
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply tests/tools/QueueCache.cc:261]: :D

File: tests/tools/QueueCache.cc [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+
+SCENARIO("Queue Caches can be subscripted", "[tools]") {
+  // Test whether the subscript operator can get values from cache.
+  GIVEN("a queue cache") {
+    const size_t capacity = 3;
+    emp::QueueCache<char, int> qch(capacity);
+
+    auto it_x = qch.Put('x', 23);
+    auto it_y = qch.Put('y', 24);
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply tests/tools/QueueCache.cc]: marked in 952e0ac

File: tests/tools/QueueCache.cc [hunk 21]
```diff
@@ -0,0 +1,261 @@
...
+    auto it_z = qch.Put('z', 25);
+
+    WHEN("we subscript into an existing key") {
+      auto res_x = qch['x'];
+      auto res_y = qch['y'];
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply tests/tools/QueueCache.cc]: done in 952e0ac

File: tests/tools/QueueCache.cc (hunk 1 above)
Comment (rodsan0) [reply tests/tools/QueueCache.cc]: made const in 952e0ac

File: tests/tools/QueueCache.cc [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+    }
+    WHEN("an element is not in it") {
+      AND_WHEN("a function isn't passed") {
+        // Test whether exception gets thrown when key is not in cache.
+        REQUIRE_THROWS_WITH(qch.Get('d'), "Key not in cache.");
+      }
+      AND_WHEN("a function is passed") {
+        THEN("it is called") {
+          // Test whether function gets called correctly when key is not found.
+          std::function<int(char)> qch_fn = [](char ch){ return (int)ch - 'a'; };
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply tests/tools/QueueCache.cc]: removed it in e231146

File: tests/tools/QueueCache.cc [hunk 23]
```diff
@@ -0,0 +1,261 @@
...
+      }
+      AND_THEN("the last element is not") {
+        REQUIRE_THROWS_WITH(qch.Get('b'), "Key not in cache.");
+      }
+    }
+  }
+}
+SCENARIO("Queue Caches can delete elements", "[tools]") {
+  GIVEN("a queue cache with elements") {
+    const size_t capacity = 2;
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply tests/tools/QueueCache.cc]: implemented in 3b6a98c

File: tests/tools/QueueCache.cc [hunk 24]
```diff
@@ -0,0 +1,261 @@
...
+
+      THEN("we get the expected result") {
+        CHECK(result == expected);
+      }
+    }
+    WHEN("we apply std::algoritms using begin() and end() iterators") {
+      emp::vector<int> result;
+      emp::vector<int> expected{'E', 'D', 'C', 'B', 'A'};
+
+      size_t distance = std::distance(
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply tests/tools/QueueCache.cc]: done in d20eca9

File: tests/tools/QueueCache.cc [hunk 25]
```diff
@@ -0,0 +1,261 @@
...
+  GIVEN("a queue cache with elements in it") {
+    const size_t capacity = 6;
+    emp::QueueCache<char, int> qch(capacity);
+
+    for (char i = 'A'; i < 'F'; ++i) {
+      qch.Put(i, i);
+    }
+
+    WHEN("we iterate on it with for loops") {
+      emp::vector<int> result;
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply tests/tools/QueueCache.cc:225]: changed in d20eca9

File: source/tools/MatchBin.h [hunk 26]
```diff
@@ -100,6 +100,8 @@ namespace emp::internal {
     #endif
     bool logging_activated;
 
+    static thread_local inline std::ofstream mystream{EMP_LOG_MATCHBIN_FILENAME};
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h]: done in 5d11fd6

File: source/hardware/EventDrivenGP.h [hunk 27]
```diff
@@ -147,7 +147,13 @@ namespace emp {
 
       Event & operator=(const Event &) = default;
       Event & operator=(Event &&) = default;
-
+      bool operator<(const Event & other) const {
```
Comment (mmore500) [inline source/hardware/EventDrivenGP.h:150]: looks like we accidentally committed some of the 53vgh hacks :/

Comment (mmore500) [reply source/hardware/EventDrivenGP.h:150]: we should get rid of these from this branch for now

File: source/tools/MatchBin.h [hunk 28]
```diff
@@ -852,25 +859,27 @@ namespace emp {
     /// Set up regulators to match target MatchBin
     void ImprintRegulators(
       const BaseMatchBin<Val, query_t, tag_t, Regulator> & target
-    ) override {
+    ) override { ImprintRegulators(target.GetState()); }
+
+    void ImprintRegulators(const state_t & target) override {
```
Comment (mmore500) [inline source/tools/MatchBin.h:864]: I think this is more 53vgh hacks :/

File: source/tools/QueueCache.h [hunk 29]
```diff
@@ -0,0 +1,192 @@
...
+
+			/// Return a constant iterator to one past the end of the cache queue.
+			/// @return Constant const_iterator to one past the end of cache queue
+			const typename cache_list_t::const_iterator cend() const {
+				return cache_list.end();
+			}
+
+			/// Return a constant iterator to the beginning of the cache queue.
+			/// Alias of cbegin()
+			/// @return Constant const_iterator to the beginning of cache queue
```
Comment (mmore500) [inline source/tools/QueueCache.h:168]: the docstrings look great!

File: source/tools/MatchBin.h (hunk 28 above)
Comment (rodsan0) [reply source/tools/MatchBin.h]: yep, fixed in b7ca8e9

File: source/hardware/EventDrivenGP.h (hunk 27 above)
Comment (rodsan0) [reply source/hardware/EventDrivenGP.h]: reverted by commit b7ca8e9

**Pull request merged**